tests/5.1/requires/test_requires_unified_shared_memory_heap_is_device_ptr.F90

```diff
@@ -0,0 +1,93 @@
+!===--- test_requires_unified_shared_memory_heap_is_device_ptr.F90 --------------------------===//
+!
+! OpenMP API Version 5.1 Aug 2021
+!
+! This test checks for unified shared memory of an array that is allocated on 
+! the heap and that is accessed from host and device with the same pointer.
+!
+! To guarantee the use of the device pointer, we use the is_device_ptr clause.
+! From Version 5.1, a list item that appears in an is_device_ptr clause must be of type C_PTR.
+!
+!===------------------------------------------------------------------------------===//
+
+#define OMPVV_MODULE_REQUIRES_LINE !$omp requires unified_shared_memory
+#include "ompvv.F90"
+
+#define N 1024
+
+PROGRAM test_requires_unified_shared_memory_heap_is_device_ptr
+   USE iso_fortran_env
+   USE iso_c_binding
+   USE ompvv_lib
+   USE omp_lib
+   implicit none
+   LOGICAL:: isOffloading
+
+!$omp requires unified_shared_memory
+
+  OMPVV_TEST_AND_SET_OFFLOADING(isOffloading)
+
+  OMPVV_WARNING_IF(.NOT. isOffloading, "With no offloading, unified shared memory is guaranteed due to host execution")
+
+   OMPVV_TEST_VERBOSE(unified_shared_memory_heap_is_device_ptr() .NE. 0)
+
+   OMPVV_REPORT_AND_RETURN()
+
+CONTAINS
+  INTEGER FUNCTION unified_shared_memory_heap_is_device_ptr()
+    INTEGER:: errors, i, ERR
+    INTEGER, TARGET, ALLOCATABLE:: anArray(:)
+    INTEGER, DIMENSION(N):: anArrayCopy
+    TYPE(C_PTR):: aPtr
+
+    OMPVV_INFOMSG("Unified shared memory testing - Array on heap")
+
+    errors = 0
+
+    ALLOCATE(anArray(N), STAT=ERR)
+
+    IF( .NOT. ALLOCATED(anArray) ) THEN
+       OMPVV_ERROR("Memory was not properly allocated")
+       OMPVV_RETURN(ERR)
+    END IF
+
+    aPtr = C_LOC(anArray(1))
+
+    DO i = 1, N
+      anArray(i) = i
+      anArrayCopy(i) = 0
+    END DO
+
+    ! Modify in the device
+    !$omp target is_device_ptr(aPtr)
+    DO i = 1, N
+      aPtr(i) = aPtr(i) + 10
+    END DO
+    !$omp end target
```

Comment:
`aPtr` is `type(c_ptr)` – and can not be accessed as array. How about:

```suggestion
    !$omp target is_device_ptr(aPtr)
    BLOCK
      integer, pointer :: anPtr(:)
      call c_f_pointer(aPtr, anPtr, [N])
      DO i = 1, N
        anPtr(i) = anPtr(i) + 10
      END DO
    END BLOCK  ! note: no '!$omp end target' as this is a 'strictly structured block'
```
which declares a local pointer variable.

_Regarding the `is_device_ptr` failure in GCC, that's a missing OpenMP 5.1 feature, tracked at https://gcc.gnu.org/PR105318_

Reply:
Thank you for catching this!